### PR TITLE
allow null selection in ChoiceFilter

### LINKED ISSUE #991: ChoiceFilter should handle NULL value (and have a better support of ALL value).
## Feature Request

This is the `ChoiceFilter` code
```php
if (\is_array($data['value'])) {
    if (0 === \count($data['value'])) {
        return;
    }

    if (\in_array('all', $data['value'], true)) {
        return;
    }

    // Have to pass IN array value as parameter. See: http://www.doctrine-project.org/jira/browse/DDC-3759
    $completeField = sprintf('%s.%s', $alias, $field);
    $parameterName = $this->getNewParameterName($queryBuilder);
    if (ChoiceType::TYPE_NOT_CONTAINS === $data['type']) {
        $this->applyWhere($queryBuilder, $queryBuilder->expr()->notIn($completeField, ':'.$parameterName));
    } else {
        $this->applyWhere($queryBuilder, $queryBuilder->expr()->in($completeField, ':'.$parameterName));
    }
    $queryBuilder->setParameter($parameterName, $data['value']);
} else {
    if ('' === $data['value'] || null === $data['value'] || false === $data['value'] || 'all' === $data['value']) {
        return;
    }

    $parameterName = $this->getNewParameterName($queryBuilder);

    if (ChoiceType::TYPE_NOT_CONTAINS === $data['type']) {
        $this->applyWhere($queryBuilder, sprintf('%s.%s <> :%s', $alias, $field, $parameterName));
    } else {
        $this->applyWhere($queryBuilder, sprintf('%s.%s = :%s', $alias, $field, $parameterName));
    }

    $queryBuilder->setParameter($parameterName, $data['value']);
}
```

You can see that:
- There is a specific behaviour for the `all` value. AFIK, this is not documented. And if you chose the `all` value, the filter is not applied, even if you chose the `NOT CONTAINS` operator. But, for me it should return no results instead of every result. IMHO, the `all` behaviour should be deprecated or at least fixed.
- I have multiple filter where sometimes I want to get all entities with the value A, sometimes all entities with the value B, and sometimes all entities with no value in database since my field is nullable. Actually the choice filter does not handle null values. 

So if `$data['value'] = null`, I expect the query to do `IS NULL` or `IS NOT NULL`.
And if `$data[`value`]` is an array with a null value, I expect to add a `OR IS NULL` check.

I can make the PR if you agree with the changes.

### PR DESCRIPTION
**Pull request by rgrassian**
## Subject

I am targeting this branch, because BC.

Closes #991

## Changelog

```markdown
### Added
- allow null to be selected in ChoiceFilter
```

### CODE REVIEW

File: src/Filter/ChoiceFilter.php [hunk 1]
```diff
@@ -35,29 +35,57 @@ public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
                 return;
             }
 
+            $isNullSelected = \in_array(null, $data['value'], true);
+            $data['value'] = array_filter($data['value'], static function ($value) {
+                return null !== $value;
+            });
```
Comment (peter-gribanov) [inline src/Filter/ChoiceFilter.php]: I think that here you can get empty array.

File: src/Filter/ChoiceFilter.php [hunk 2]
```diff
@@ -35,29 +35,57 @@ public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
...
+                    ));
+                    $queryBuilder->setParameter($parameterName, $data['value']);
+                }
             } else {
-                $this->applyWhere($queryBuilder, $queryBuilder->expr()->in($completeField, ':'.$parameterName));
+                if ($isNullSelected) {
+                    $this->applyWhere($queryBuilder, $queryBuilder->expr()->orX(
+                        $queryBuilder->expr()->in($completeField, ':'.$parameterName),
+                        $queryBuilder->expr()->isNull($completeField)
+                    ));
```
Comment (peter-gribanov) [inline src/Filter/ChoiceFilter.php]: Empty data array will result in an incorrect query here.
```sql
SELECT * FROM ... WHERE field IN () OR field IS NULL
```

Comment (VincentLanglet) [reply src/Filter/ChoiceFilter.php]: Empty data array will no result in an incorrect query.
Empty data are always replaced by `[null]`.

And we get the following query
```
SELECT * FROM ... WHERE field IN (NULL) OR field IS NULL
```

File: src/Filter/ChoiceFilter.php (hunk 1 above)
Comment (VincentLanglet) [reply src/Filter/ChoiceFilter.php]: Yes, but this is not a problem.

**Comment by peter-gribanov** [comment]
IMHO The code turned out to be rather cumbersome and difficult to understand. Maybe you should apply the strategy pattern or separate the logic by methods?

File: src/Filter/ChoiceFilter.php (hunk 2 above)
Comment (peter-gribanov) [reply src/Filter/ChoiceFilter.php]: Yes. You're right. I'm confusing something.

**Comment by VincentLanglet** [comment]
> IMHO The code turned out to be rather cumbersome and difficult to understand. Maybe you should apply the strategy pattern or separate the logic by methods?

We improved the way the query is build. I don't really see any difficulty to understand what is done so I don't think we should create privateMethod just to create privateMethod.

**Comment by VincentLanglet** [comment]
@greg0ire @OskarStark @phansys I'll be happy if you find some time to review this PR :)

**Comment by VincentLanglet** [comment]
> Looks good to me, but I didn't test it in a real project 👍

We started by writing the code directly in our vendor in order to check this works.

**Comment by VincentLanglet** [comment]
@greg0ire @core23 @phansys If you have time to take look, it would be great.
We would like to use this new feature in our project :)

File: src/Filter/ChoiceFilter.php [hunk 3]
```diff
@@ -35,29 +35,51 @@ public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
...
+            $data['value'] = array_filter($data['value'], static function ($value) {
+                return null !== $value;
+            });
+
             // Have to pass IN array value as parameter. See: http://www.doctrine-project.org/jira/browse/DDC-3759
             $completeField = sprintf('%s.%s', $alias, $field);
             $parameterName = $this->getNewParameterName($queryBuilder);
             if (ChoiceType::TYPE_NOT_CONTAINS === $data['type']) {
-                $this->applyWhere($queryBuilder, $queryBuilder->expr()->notIn($completeField, ':'.$parameterName));
+                $andConditions = [$queryBuilder->expr()->isNotNull($completeField)];
```
Comment (core23) [inline src/Filter/ChoiceFilter.php]: This method is very long. Can you please try to extract some small methods. This improves reliability.

Comment (VincentLanglet) [reply src/Filter/ChoiceFilter.php]: Does 
```
if (\is_array($data['value'])) {
    $this->filterMultipleValues(...)
else {
    $this->filterSimpleValue()
}
```

would be enough ? @core23 @peter-gribanov

File: src/Filter/ChoiceFilter.php [hunk 4]
```diff
@@ -35,29 +35,51 @@ public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
...
             }
 
             $parameterName = $this->getNewParameterName($queryBuilder);
 
             if (ChoiceType::TYPE_NOT_CONTAINS === $data['type']) {
-                $this->applyWhere($queryBuilder, sprintf('%s.%s <> :%s', $alias, $field, $parameterName));
+                if (null === $data['value']) {
+                    $this->applyWhere($queryBuilder, $queryBuilder->expr()->isNotNull(sprintf('%s.%s', $alias, $field)));
+                } else {
+                    $this->applyWhere($queryBuilder, sprintf('%s.%s <> :%s', $alias, $field, $parameterName));
```
Comment (phansys) [inline src/Filter/ChoiceFilter.php]: I'd use `!=` or `Expr::neq()` instead.

File: tests/Filter/ChoiceFilterTest.php [hunk 5]
```diff
@@ -55,11 +55,35 @@ public function testFilterArray(): void
 
         $filter->filter($builder, 'alias', 'field', ['type' => ChoiceType::TYPE_CONTAINS, 'value' => ['1', '2']]);
 
-        $this->assertSame(['in_alias.field', 'in_alias.field IN :field_name_0'], $builder->query);
+        $this->assertSame('in_alias.field', $builder->query[0]);
+        $this->assertSame('in_alias.field IN :field_name_0', $builder->query[1]->__toString());
```
Comment (phansys) [inline tests/Filter/ChoiceFilterTest.php]:
```suggestion
        $this->assertSame('in_alias.field IN :field_name_0', (string) $builder->query[1]);
```

File: tests/Filter/ChoiceFilterTest.php [hunk 6]
```diff
@@ -55,11 +55,35 @@ public function testFilterArray(): void
...
+    {
+        $filter = new ChoiceFilter();
+        $filter->initialize('field_name', ['field_options' => ['class' => 'FooBar']]);
+
+        $builder = new ProxyQuery(new QueryBuilder());
+
+        $filter->filter($builder, 'alias', 'field', ['type' => ChoiceType::TYPE_CONTAINS, 'value' => ['1', null]]);
+
+        $this->assertSame('in_alias.field', $builder->query[0]);
+        $this->assertSame('in_alias.field IN :field_name_0 OR alias.field IS NULL', $builder->query[1]->__toString());
```
Comment (phansys) [inline tests/Filter/ChoiceFilterTest.php]:
```suggestion
        $this->assertSame('in_alias.field IN :field_name_0 OR alias.field IS NULL', (string) $builder->query[1]);
```

File: tests/Filter/ChoiceFilterTest.php [hunk 7]
```diff
@@ -55,11 +55,35 @@ public function testFilterArray(): void
...
+        $this->assertSame(['field_name_0' => ['1']], $builder->parameters);
+        $this->assertTrue($filter->isActive());
+
+        $builder = new ProxyQuery(new QueryBuilder());
+
+        $filter->filter($builder, 'alias', 'field', ['type' => ChoiceType::TYPE_NOT_CONTAINS, 'value' => ['1', null]]);
+
+        $this->assertSame('alias.field IS NOT NULL AND alias.field NOT IN :field_name_0', $builder->query[0]->__toString());
```
Comment (phansys) [inline tests/Filter/ChoiceFilterTest.php]:
```suggestion
        $this->assertSame('alias.field IS NOT NULL AND alias.field NOT IN :field_name_0', (string) $builder->query[0]);
```

File: tests/Filter/QueryBuilder.php [hunk 8]
```diff
@@ -93,6 +94,14 @@ public function orX($x = null)
         return new Orx(\func_get_args());
     }
 
+    /**
+     * @return Andx
+     */
+    public function andX($x = null)
```
Comment (phansys) [inline tests/Filter/QueryBuilder.php]:
```suggestion
    public function andX($x = null): Andx
```

File: tests/Filter/QueryBuilder.php [hunk 9]
```diff
@@ -114,6 +123,16 @@ public function isNull($queryPart)
         return $queryPart.' IS NULL';
     }
 
+    /**
+     * @param string $queryPart
+     *
+     * @return string
+     */
+    public function isNotNull($queryPart)
```
Comment (phansys) [inline tests/Filter/QueryBuilder.php]:
```suggestion
    public function isNotNull(string $queryPart): string
```

File: src/Filter/ChoiceFilter.php [hunk 10]
```diff
@@ -79,4 +51,64 @@ public function getRenderSettings()
             'label' => $this->getLabel(),
         ]];
     }
+
+    private function filterWithMultipleValues(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
```
Comment (phansys) [inline src/Filter/ChoiceFilter.php]:
```suggestion
    private function filterWithMultipleValues(ProxyQueryInterface $queryBuilder, string $alias, string $field, array $data = []): void
```

File: src/Filter/ChoiceFilter.php [hunk 11]
```diff
@@ -79,4 +51,64 @@ public function getRenderSettings()
...
+            $orConditions = [$queryBuilder->expr()->in($completeField, ':'.$parameterName)];
+            if ($isNullSelected) {
+                $orConditions[] = $queryBuilder->expr()->isNull($completeField);
+            }
+            $this->applyWhere($queryBuilder, $queryBuilder->expr()->orX()->addMultiple($orConditions));
+            $queryBuilder->setParameter($parameterName, $data['value']);
+        }
+    }
+
+    private function filterWithSingleValue(ProxyQueryInterface $queryBuilder, $alias, $field, $data)
```
Comment (phansys) [inline src/Filter/ChoiceFilter.php]:
```suggestion
    private function filterWithSingleValue(ProxyQueryInterface $queryBuilder, string $alias, string $field, array $data = []): void
```

File: tests/Filter/QueryBuilder.php [hunk 12]
```diff
@@ -114,6 +120,14 @@ public function isNull($queryPart)
         return $queryPart.' IS NULL';
     }
 
+    /**
+     * @param string $queryPart
+     */
+    public function isNotNull($queryPart): string
```
Comment (phansys) [inline tests/Filter/QueryBuilder.php]:
```suggestion
    public function isNotNull(string $queryPart): string
```

**Comment by rgrassian** [comment]
@core23 and @phansys it's done

**Comment by VincentLanglet** [comment]
@core23 Can you take a look ? I wait for you approbation to merge this :)

**Comment by VincentLanglet** [comment]
Thanks !

**Comment by VincentLanglet** [comment]
@greg0ire Can you make a new release ? :)

**Pull request merged**